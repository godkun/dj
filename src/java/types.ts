import { JavaClazz } from './core/clazz'

export interface IType {
  tsType: string
  javaType: string
  generic?: Array<JavaClazz | IType>
}

export interface IJavaClazzField {
  name: string
  type: IType
  comment?: string
}

export type TServiceMethod = {
  [k in string]: IMethodMeta
}

export type IMethodArgs = {
  name: string
  type: string
}

export interface IMethodMeta {
  args: Array<IMethodArgs>
  ret: string
}
