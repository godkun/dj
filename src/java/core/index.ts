/**
 * 通过构造Java Style code的DSL来自动生成 ts，java， go的dubbo框架的调用代码
 */

import Clazz from './clazz'
import Service from './service'
import {
  Integer,
  String,
  Boolean,
  List,
  Set,
  Collection,
  Map,
  boolean,
  int,
  short,
  Short,
  byte,
  Byte,
  long,
  Long,
  double,
  Double,
  float,
  Float,
  char,
  Char,
  chars,
  Character,
  Enumeration,
  Iterator,
  Currency,
  Dictionary,
  HashMap,
} from './java-type'

import Enum from './enum'

export default {
  Enum,
  Clazz,
  Service,

  Boolean,
  boolean,

  Integer,
  int,

  short,
  Short,

  byte,
  Byte,

  long,
  Long,

  double,
  Double,

  float,
  Float,

  String,
  char,
  Char,
  chars,
  Character,

  List,
  Set,
  Collection,
  Iterator,
  Enumeration,
  Map,
  HashMap,
  Dictionary,
  Currency,
}
