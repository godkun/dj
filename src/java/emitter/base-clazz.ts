import path from 'path'

import debug from 'debug'
import Deps from '../core/deps'
import fsx from 'fs-extra'
import fmt from './fmt'

const log = debug(`dubbo:dj:emitter/BaseClazzï¼šsay ~`)
const relWhiteList = ['js-to-java', '@dubbo/dj-suger']

const fileType = {
  ts: 'ts',
  go: 'go',
  java: 'java',
}

export default abstract class BaseClazz {
  curFilePath: string

  constructor(
    public fullClsName: string,
    lang: 'ts' | 'go' | 'java',
    baseDir: string
  ) {
    this.curFilePath = `${baseDir}/${lang}/${this.cls2Path(fullClsName)}.${
      fileType[lang]
    }`
  }

  imports(deps: Deps) {
    const imports = []
    for (let { fullClassName, importName } of deps.imports) {
      if (
        relWhiteList.includes(fullClassName) ||
        fullClassName.startsWith('.')
      ) {
        imports.push(`import ${importName} from '${fullClassName}'`)
      } else {
        let rel = this.relPath(this.fullClsName, fullClassName)
        if (!rel.startsWith('./') && !rel.startsWith('../')) {
          rel = './' + rel
        }
        imports.push(`import ${importName} from '${rel}'`)
      }
    }
    return imports
  }

  get mott() {
    return `
      /**
       * auto generated by dubbo dj
       * ~~~ ðŸ’— mechine coding ðŸ’— ~~~
       */
    `
  }

  cls2Path(cls: string) {
    return cls.replace(/\./g, '/')
  }

  relPath(from: string, to: string) {
    const fromArr = from.split('.')
    fromArr.pop()
    // è®¡ç®—ç›¸å¯¹è·¯å¾„
    const relPath = path.relative(fromArr.join('/'), this.cls2Path(to))
    log(`${from} => relative => ${to} =>`, relPath)
    return relPath
  }

  // write to => ./dubbo/ts/**/*.ts
  async writeCode(): Promise<void> {
    try {
      await fsx.ensureFile(this.curFilePath)
      await fsx.writeFile(this.curFilePath, this.pretteyCode)
      return log(
        `create ${this.fullClsName} => ${this.curFilePath} was successfully`
      )
    } catch (err) {
      log(`generate => '${this.fullClsName}' was failed, err:` + err)
      console.error(err)
    }
  }

  get pretteyCode(): string {
    return fmt(this.code)
  }

  abstract get code(): string
}
