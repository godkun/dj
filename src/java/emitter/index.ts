import Clazz from './clazz'
import Enum from './enum'
import Service from './service'
import AbstractService from './service-abstract'

export { Enum, Clazz, Service, AbstractService }
