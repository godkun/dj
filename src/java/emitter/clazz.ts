import JavaBase from './base-clazz'
import { JavaClazz } from '../core/clazz'
import debug from 'debug'
import { clazzDot } from './dot/ts'

const log = debug(`dubbo:dj:emitter/clazzï¼šsay ~`)

export default class Clazz extends JavaBase {
  constructor(
    private clazz: JavaClazz,
    lang: 'ts' | 'go' | 'java',
    private baseDir: string = './dubbo'
  ) {
    super(clazz.fullClsName, lang, baseDir)
  }

  get code(): string {
    return clazzDot({
      mott: this.mott,
      imports: this.imports(this.clazz.deps).join(';'),
      cls: this.clazz.fullClsName,
      infName: this.clazz.infName,
      fields: this.clazz.fields,
      clsName: this.clazz.clsName,
    })
  }
}
