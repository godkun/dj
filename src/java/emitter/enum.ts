import { JavaEnum } from '../core/enum'
import JavaBase from './base-clazz'
import { enumDot } from './dot/ts'

export default class Enum extends JavaBase {
  constructor(
    public javaEnum: JavaEnum,
    lang: 'ts' | 'go' | 'java',
    baseDir: string = './dubbo'
  ) {
    super(javaEnum.fullClsName, lang, baseDir)
  }

  get code() {
    return enumDot({
      mott: this.mott,
      name: this.javaEnum.clsName,
      fields: this.javaEnum.fields,
    })
  }
}
