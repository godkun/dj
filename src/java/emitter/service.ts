import fsx, { exists } from 'fs-extra'
import JavaBase from './base-clazz'
import { JavaService } from '../core/service'
import { serviceDot } from './dot/ts'
import debug from 'debug'

const log = debug(`dubbo:dj:service:say ~`)

export default class Service extends JavaBase {
  constructor(
    public service: JavaService,
    lang: 'ts' | 'go' | 'java',
    baseDir: string = './dubbo'
  ) {
    super(service.fullClsName, lang, baseDir)
  }

  async writeCode(): Promise<void> {
    // 如果存在就不覆盖
    const exists = await fsx.pathExists(this.curFilePath)
    if (!exists) {
      super.writeCode()
    } else {
      log(`${this.curFilePath} was existed, skip it ~~~~~~`)
    }
  }

  get code() {
    const methods = []
    const args = []

    for (let [name, meta] of Object.entries(this.service._methods)) {
      for (let arg of meta.args) {
        args.push(`${arg.name}: ${arg.type}`)
      }
      methods.push(`
        ${name}(${args.join()})${meta.ret ? ':' + meta.ret : 'void'} {
          throw new Error('Method not implemented.')
        }
      `)
    }

    return serviceDot({
      mott: this.mott,
      imports: this.imports(this.service.deps).join(';'),
      service: this.service.fullClsName,
      serviceName: this.service.clsName,
      group: this.service._group,
      version: this.service._version,
      methods: methods,
    })
  }
}
