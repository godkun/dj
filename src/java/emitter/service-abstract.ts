import JavaBase from './base-clazz'
import { JavaService } from '../core/service'
import { abstractServiceDot } from './dot/ts'

export default class AbstractService extends JavaBase {
  constructor(
    public service: JavaService,
    lang: 'ts' | 'go' | 'java',
    baseDir: string = './dubbo'
  ) {
    super(service.fullClsName, lang, baseDir)
    this.curFilePath = this.curFilePath.replace(
      service.clsName,
      `base/${service.clsName}`
    )
    this.fullClsName = this.fullClsName.replace(
      service.clsName,
      `base.${service.clsName}`
    )
  }

  get code() {
    const methods = []
    const args = []

    for (let [name, meta] of Object.entries(this.service._methods)) {
      for (let arg of meta.args) {
        args.push(`${arg.name}: ${arg.type}`)
      }
      methods.push(
        `abstract ${name}(${args.join()})${meta.ret ? ':' + meta.ret : 'void'};`
      )
    }

    return abstractServiceDot({
      mott: this.mott,
      imports: this.imports(this.service.deps).join(';'),
      service: this.service.fullClsName,
      serviceName: this.service.clsName,
      group: this.service._group,
      version: this.service._version,
      methods: methods,
    })
  }
}
