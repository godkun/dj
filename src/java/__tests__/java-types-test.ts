import java from '../core'
import { Clazz, Enum } from '../emitter'

it('test java list', () => {
  const color = java
    .Enum('org.apache.dubbo.service.Color')
    .field('Red', 0)
    .field('Green', 1)
    .field('Blue', 2)
    .ok()

  const foo = java
    .Clazz('org.apache.dubbo.domain.Foo')
    .field('id', java.Integer)
    .ok()

  const user = java
    .Clazz('org.apache.duboo.domain.User')
    .field('id', java.Boolean)
    .field('likes', java.List(java.String))
    .field('list', java.List(foo))
    .field('listFoo', java.List(foo))
    .field('fooSet', java.Set(foo))
    .field('fooCollection', java.Collection(foo))
    .field('map', java.Map(java.String, java.String))
    .ok()

  expect(new Clazz(user, 'ts').pretteyCode).toMatchSnapshot()
  expect(new Enum(color, 'ts')).toMatchSnapshot()
})
