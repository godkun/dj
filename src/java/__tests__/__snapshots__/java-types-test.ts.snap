// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test java list 1`] = `
"/**
 * auto generated by dubbo dj
 * ~~~ ðŸ’— mechine coding ðŸ’— ~~~
 */

import java from 'js-to-java'
import s from '@dubbo/dj-suger'
import Foo from '../../dubbo/domain/Foo'

export interface IUser {
  id: boolean
  likes: Array<string>
  list: Array<Foo>
  listFoo: Array<Foo>
  fooSet: Array<Foo>
  fooCollection: Array<Foo>
  map: Map<string, string>
}

export default class User {
  id: boolean
  likes: Array<string>
  list: Array<Foo>
  listFoo: Array<Foo>
  fooSet: Array<Foo>
  fooCollection: Array<Foo>
  map: Map<string, string>

  constructor(props: IUser) {
    this.id = props.id
    this.likes = props.likes
    this.list = props.list
    this.listFoo = props.listFoo
    this.fooSet = props.fooSet
    this.fooCollection = props.fooCollection
    this.map = props.map
  }

  __fields2java() {
    java('org.apache.duboo.domain.User', {
      id: java.Boolean(this.id),
      likes: java.List(s.$lhs(this.likes, java.String)),
      list: java.List(s.$lhs(this.list)),
      listFoo: java.List(s.$lhs(this.listFoo)),
      fooSet: java.Set(s.$lhs(this.fooSet)),
      fooCollection: java.Collection(s.$lhs(this.fooCollection)),
      map: s.$mhs(this.map, java.String),
    })
  }
}
"
`;

exports[`test java list 2`] = `
Enum {
  "curFilePath": "./dubbo/ts/org/apache/dubbo/service/Color.ts",
  "fullClsName": "org.apache.dubbo.service.Color",
  "javaEnum": JavaEnum {
    "clsName": "Color",
    "deps": Deps {
      "map": Map {},
    },
    "fields": Array [
      Object {
        "comment": "",
        "name": "Red",
        "val": 0,
      },
      Object {
        "comment": "",
        "name": "Green",
        "val": 1,
      },
      Object {
        "comment": "",
        "name": "Blue",
        "val": 2,
      },
    ],
    "fullClsName": "org.apache.dubbo.service.Color",
    "infName": "IColor",
  },
}
`;
